orderbook/src/account_balances.rs

```diff
@@ -45,73 +56,74 @@ pub struct Web3BalanceFetcher {
 struct SubscriptionKey {
     owner: H160,
     token: H160,
+    source: SellTokenSource,
 }
 
-#[derive(Clone, Default)]
+#[derive(Clone, Debug, Default)]
 struct SubscriptionValue {
     balance: Option<U256>,
     allowance: Option<U256>,
 }
 
 impl Web3BalanceFetcher {
-    pub fn new(web3: Web3, vault_relayer: H160, settlement_contract: H160) -> Self {
+    pub fn new(
+        web3: Web3,
+        vault: Option<BalancerV2Vault>,
+        vault_relayer: H160,
+        settlement_contract: H160,
+    ) -> Self {
         Self {
             web3,
+            vault,
             vault_relayer,
             settlement_contract,
             balances: Default::default(),
         }
     }
 
-    async fn _register_many(
-        &self,
-        subscriptions: impl Iterator<Item = SubscriptionKey>,
-    ) -> Result<()> {
+    async fn _register_many(&self, subscriptions: Vec<SubscriptionKey>) -> Result<()> {
         let mut batch = CallBatch::new(self.web3.transport());
 
         // Make sure subscriptions are registered for next update even if batch call fails
-        let subscriptions: Vec<SubscriptionKey> = {
+        {
             let mut guard = self.balances.lock().expect("thread holding mutex panicked");
-            subscriptions
-                .map(|subscription| {
-                    let _ = guard.entry(subscription).or_default();
-                    subscription
-                })
-                .collect()
-        };
+            for subscription in &subscriptions {
+                guard.insert(*subscription, Default::default());
+            }
```

Comment:
This is a logic change. Before we were making sure there is at least the default but now we are always overwriting with the default. This means that while the fetching happens below the balance and allowance are 0 instead of retaining their previous value. Not sure if this is intended.
Either way we could add a comment here explaining which behavior we want.
Reading this now I'm not sure why we need default values in the first place. Does this prevent duplicate work when getting balances simultaneously somewhere?

Reply:
Nice catch!

> Reading this now I'm not sure why we need default values in the first place. Does this prevent duplicate work when getting balances simultaneously somewhere?

There might be a subtle difference where we return `None` while updating the account balance where we used to return the previous account balance.

I will change this back to the original and add a comment.

Reply:
Should be fixed now.